mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/DecoratedMessager.java

```diff
@@ -37,135 +39,93 @@ public void warn(String template, Object... arguments) {
-            this.locationInfo = "";
-          } else {
-            this.actualTargetElement = processedType;
-            this.locationInfo =
-                String.format(
-                    "[Parameter %s of %s inherited from %s] ",
-                    intendedTargetElement.getSimpleName(),
-                    method.getSimpleName(),
-                    declaringType.getSimpleName());
-          }
-          break;
-        default:
-          // We don't emit messages for other types of elements in the mapper processor. Handle
-          // gracefully nevertheless:
-          this.actualTargetElement = intendedTargetElement;
-          this.locationInfo = "";
-          break;
-      }
+  private void message(
+      Diagnostic.Kind level, Element element, String template, Object[] arguments) {
+    if (emittedMessages.add(new MessageId(level, element, template, arguments))) {
+      messager.printMessage(
+          level, formatLocation(element) + String.format(template, arguments), element);
     }
+  }
 
-    void print(Diagnostic.Kind level, String template, Object... arguments) {
-      messager.printMessage(
-          level, String.format(locationInfo + template, arguments), actualTargetElement);
+  private static String formatLocation(Element element) {
+    switch (element.getKind()) {
+      case CLASS:
+      case INTERFACE:
+        return String.format("[%s] ", element.getSimpleName());
+      case FIELD:
+      case METHOD:
+      case CONSTRUCTOR:
+        return String.format(
+            "[%s.%s] ", element.getEnclosingElement().getSimpleName(), element.getSimpleName());
+      case PARAMETER:
+        Element method = element.getEnclosingElement();
+        Element type = method.getEnclosingElement();
+        return String.format(
+            "[%s.%s, parameter %s] ",
+            type.getSimpleName(), method.getSimpleName(), element.getSimpleName());
+      default:
+        // We don't emit messages for other types of elements in the mapper processor. Handle
+        // gracefully nevertheless:
+        return String.format("[%s] ", element);
+    }
+  }
+
+  private static class MessageId {
+
+    private final Diagnostic.Kind level;
+    private final Element element;
+    private final String template;
+    private final Object[] arguments;
+
+    private MessageId(Diagnostic.Kind level, Element element, String template, Object[] arguments) {
+      this.level = level;
+      this.element = element;
+      this.template = template;
+      this.arguments = arguments;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other == this) {
+        return true;
+      } else if (other instanceof MessageId) {
+        MessageId that = (MessageId) other;
+        return this.level == that.level
+            && Objects.equals(this.element, that.element)
+            && Objects.equals(this.template, that.template)
+            && Arrays.deepEquals(this.arguments, that.arguments);
```

Comment:
will this `deepEquals` give us expected behavior? Maybe it would be better to convert the 
`Object[] arguments ` to `List` and do `Objects.equals()` on it?

Reply:
Why wouldn't `deepEquals` give "expected behavior"? Its contract is very well defined:

> Two array references are considered deeply equal if both are null, or if they refer to arrays that contain the same number of elements and all corresponding pairs of elements in the two arrays are deeply equal.